### PR TITLE
Introduce MemTable

### PR DESCRIPTION
**Pull request by JingsongLi**
MemTable
- Append only memory table for storing key-values.
- When it is full, it will be flushed to disk and form an SST file.

### CODE REVIEW

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/SortBufferMemTableTest.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        for (int i = 0; i < results.size(); i++) {
+            Assert.assertTrue(isEquals(inputs.get(i), results.get(i)));
+        }
+    }
+
+    private boolean isEquals(KeyValue kv1, KeyValue kv2) {
+        return kv1.sequenceNumber() == kv2.sequenceNumber()
+                && kv1.valueKind() == kv2.valueKind()
+                && kv1.key().getInt(0) == kv2.key().getInt(0)
+                && kv1.value().getInt(0) == kv2.value().getInt(0);
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/SortBufferMemTableTest.java:137]: Why not add a `hashCode` and `equals` method to `KeyValue`. Other tests can also use them.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/HeapMemorySegmentPool.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.table.runtime.util.MemorySegmentPool;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/** MemorySegment pool from heap. */
+public class HeapMemorySegmentPool implements MemorySegmentPool {
```
Comment (tsreaper) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/HeapMemorySegmentPool.java:29]: Why do we need a new memory segment pool? Isn't `LazyMemorySegmentPool` sufficient?

Comment (JingsongLi) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/utils/HeapMemorySegmentPool.java:29]: Sink does not support getting memory from `MemoryManager` now.

File: flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/SortBufferMemTableTest.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/SortBufferMemTableTest.java:137]: For this case, the expected row is `GenericRowData`, but the row in returned `KeyValue` is `OffsetRowData`. They can not be compared.

**Review by tsreaper — APPROVED** [review]

**Pull request merged**